File: libs/wasm-sdk/wasm-sdk-swift/Sources/FlowLikeSDK/JSON.swift
// JSON.swift — Hand-rolled JSON utilities (no Foundation).

// MARK: - JSON Escape

public func jsonEscape(_ s: String) -> String {
    var result: [UInt8] = []
    result.reserveCapacity(s.utf8.count)
    for c in s.utf8 {
        switch c {
        case 0x22: // "
            result.append(0x5C); result.append(0x22)
        case 0x5C: // backslash
            result.append(0x5C); result.append(0x5C)
        case 0x0A: // \n
            result.append(0x5C); result.append(0x6E)
        case 0x0D: // \r
            result.append(0x5C); result.append(0x72)
        case 0x09: // \t
            result.append(0x5C); result.append(0x74)
        default:
            result.append(c)
        }
    }
    return String(_bytes: result)
}

public func jsonQuote(_ s: String) -> String {
    "\"" + jsonEscape(s) + "\""
}

// MARK: - String from UTF-8 bytes (no Foundation)

extension String {
    init(_bytes: [UInt8]) {
        self = _bytes.withUnsafeBufferPointer { buf in
            String(decoding: buf, as: UTF8.self)
        }
    }
}

// MARK: - JSONBuilder

public struct JSONBuilder: ~Copyable, Sendable {
    var buffer: [UInt8] = []
    var needsComma: Bool = false

    public init() {}

    public mutating func beginObject() {
        if needsComma { buffer.append(0x2C) } // ,
        buffer.append(0x7B) // {
        needsComma = false
    }

    public mutating func endObject() {
        buffer.append(0x7D) // }
        needsComma = true
    }

    public mutating func beginArray() {
        if needsComma { buffer.append(0x2C) } // ,
        buffer.append(0x5B) // [
        needsComma = false
    }

    public mutating func endArray() {
        buffer.append(0x5D) // ]
        needsComma = true
    }

    public mutating func addComma() {
        buffer.append(0x2C) // ,
    }

    public mutating func addKey(_ key: String) {
        if needsComma { buffer.append(0x2C) }
        appendQuotedString(key)
        buffer.append(0x3A) // :
        needsComma = false
    }

    public mutating func addKeyRaw(_ key: String) {
        appendQuotedString(key)
        buffer.append(0x3A) // :
        needsComma = false
    }

    public mutating func addString(_ value: String) {
        appendQuotedString(value)
        needsComma = true
    }

    public mutating func addInt(_ value: Int) {
        appendIntLiteral(value)
        needsComma = true
    }

    public mutating func addBool(_ value: Bool) {
        if value {
            for c in "true".utf8 { buffer.append(c) }
        } else {
            for c in "false".utf8 { buffer.append(c) }
        }
        needsComma = true
    }

    public mutating func addRaw(_ raw: String) {
        for c in raw.utf8 { buffer.append(c) }
        needsComma = true
    }

    public consuming func build() -> String {
        String(_bytes: buffer)
    }

    // MARK: - Internal helpers

    mutating func appendQuotedString(_ s: String) {
        buffer.append(0x22) // "
        for c in s.utf8 {
            switch c {
            case 0x22:
                buffer.append(0x5C); buffer.append(0x22)
            case 0x5C:
                buffer.append(0x5C); buffer.append(0x5C)
            case 0x0A:
                buffer.append(0x5C); buffer.append(0x6E)
            case 0x0D:
                buffer.append(0x5C); buffer.append(0x72)
            case 0x09:
                buffer.append(0x5C); buffer.append(0x74)
            default:
                buffer.append(c)
            }
        }
        buffer.append(0x22) // "
    }

    mutating func appendIntLiteral(_ value: Int) {
        if value == 0 {
            buffer.append(0x30) // '0'
            return
        }
        var v = value
        var negative = false
        if v < 0 {
            negative = true
            v = -v
        }
        var digits: [UInt8] = []
        while v > 0 {
            digits.append(UInt8(v % 10) + 0x30)
            v /= 10
        }
        if negative { buffer.append(0x2D) } // '-'
        for d in digits.reversed() { buffer.append(d) }
    }
}

// MARK: - Minimal JSON parser for ExecutionInput

public func parseExecutionInput(_ s: String) -> ExecutionInput {
    var input = ExecutionInput()
    let bytes = Array(s.utf8)
    var idx = 0

    func skipWhitespace() {
        while idx < bytes.count {
            let c = bytes[idx]
            if c == 0x20 || c == 0x09 || c == 0x0A || c == 0x0D {
                idx += 1
            } else {
                break
            }
        }
    }

    func readString() -> String {
        guard idx < bytes.count, bytes[idx] == 0x22 else { return "" }
        idx += 1 // skip opening "
        var result: [UInt8] = []
        while idx < bytes.count && bytes[idx] != 0x22 {
            if bytes[idx] == 0x5C { // backslash
                idx += 1
                if idx < bytes.count {
                    switch bytes[idx] {
                    case 0x6E: result.append(0x0A)
                    case 0x72: result.append(0x0D)
                    case 0x74: result.append(0x09)
                    case 0x22: result.append(0x22)
                    case 0x5C: result.append(0x5C)
                    default: result.append(bytes[idx])
                    }
                    idx += 1
                }
            } else {
                result.append(bytes[idx])
                idx += 1
            }
        }
        if idx < bytes.count { idx += 1 } // skip closing "
        return String(_bytes: result)
    }

    func readValue() -> String {
        skipWhitespace()
        guard idx < bytes.count else { return "" }
        let c = bytes[idx]
        if c == 0x22 { // "
            let v = readString()
            return "\"" + jsonEscape(v) + "\""
        } else if c == 0x7B { // {
            var depth = 0
            let start = idx
            while idx < bytes.count {
                if bytes[idx] == 0x7B { depth += 1 }
                else if bytes[idx] == 0x7D {
                    depth -= 1
                    if depth == 0 { idx += 1; return String(_bytes: Array(bytes[start..<idx])) }
                } else if bytes[idx] == 0x22 {
                    idx += 1
                    while idx < bytes.count && bytes[idx] != 0x22 {
                        if bytes[idx] == 0x5C { idx += 1 }
                        idx += 1
                    }
                }
                idx += 1
            }
            return String(_bytes: Array(bytes[start..<idx]))
        } else if c == 0x5B { // [
            var depth = 0
            let start = idx
            while idx < bytes.count {
                if bytes[idx] == 0x5B { depth += 1 }
                else if bytes[idx] == 0x5D {
                    depth -= 1
                    if depth == 0 { idx += 1; return String(_bytes: Array(bytes[start..<idx])) }
                } else if bytes[idx] == 0x22 {
                    idx += 1
                    while idx < bytes.count && bytes[idx] != 0x22 {
                        if bytes[idx] == 0x5C { idx += 1 }
                        idx += 1
                    }
                }
                idx += 1
            }
            return String(_bytes: Array(bytes[start..<idx]))
        } else {
            let start = idx
            while idx < bytes.count {
                let b = bytes[idx]
                if b == 0x2C || b == 0x7D || b == 0x5D ||
                   b == 0x20 || b == 0x09 || b == 0x0A || b == 0x0D {
                    break
                }
                idx += 1
            }
            return String(_bytes: Array(bytes[start..<idx]))
        }
    }

    skipWhitespace()
    guard idx < bytes.count, bytes[idx] == 0x7B else { return input }
    idx += 1 // skip {

    while idx < bytes.count {
        skipWhitespace()
        if idx >= bytes.count || bytes[idx] == 0x7D { break }
        if bytes[idx] == 0x2C { idx += 1; continue }
        let key = readString()
        skipWhitespace()
        if idx < bytes.count && bytes[idx] == 0x3A { idx += 1 }
        skipWhitespace()

        switch key {
        case "node_id": input.nodeId = readString()
        case "node_name": input.nodeName = readString()
        case "run_id": input.runId = readString()
        case "app_id": input.appId = readString()
        case "board_id": input.boardId = readString()
        case "user_id": input.userId = readString()
        case "stream_state":
            let v = readValue()
            input.streamState = (v == "true")
        case "log_level":
            let v = readValue()
            if let digit = v.utf8.first, digit >= 0x30 && digit <= 0x39 {
                input.logLevel = digit - 0x30
            }
        case "inputs":
            skipWhitespace()
            if idx < bytes.count && bytes[idx] == 0x7B {
                idx += 1
                while idx < bytes.count {
                    skipWhitespace()
                    if idx >= bytes.count || bytes[idx] == 0x7D { idx += 1; break }
                    if bytes[idx] == 0x2C { idx += 1; continue }
                    let iKey = readString()
                    skipWhitespace()
                    if idx < bytes.count && bytes[idx] == 0x3A { idx += 1 }
                    let iVal = readValue()
                    input.inputs[iKey] = iVal
                }
            } else {
                _ = readValue()
            }
        default:
            _ = readValue()
        }
    }

    return input
}

File: libs/wasm-sdk/wasm-sdk-java/src/main/java/com/flowlike/sdk/Json.java
package com.flowlike.sdk;

import java.util.HashMap;
import java.util.Map;

public final class Json {

    private Json() {}

    public static String escape(String s) {
        if (s == null) return "";
        StringBuilder b = new StringBuilder(s.length());
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"':  b.append("\\\""); break;
                case '\\': b.append("\\\\"); break;
                case '\n': b.append("\\n");  break;
                case '\r': b.append("\\r");  break;
                case '\t': b.append("\\t");  break;
                default:   b.append(c);
            }
        }
        return b.toString();
    }

    public static String quote(String s) {
        return "\"" + escape(s) + "\"";
    }

    public static Types.ExecutionInput parseExecutionInput(String s) {
        Types.ExecutionInput input = new Types.ExecutionInput();
        if (s == null || s.isEmpty()) return input;

        int[] pos = {0};

        skipWhitespace(s, pos);
        if (pos[0] >= s.length() || s.charAt(pos[0]) != '{') return input;
        pos[0]++;

        while (pos[0] < s.length()) {
            skipWhitespace(s, pos);
            if (pos[0] >= s.length() || s.charAt(pos[0]) == '}') break;
            if (s.charAt(pos[0]) == ',') { pos[0]++; continue; }

            String key = readString(s, pos);
            skipWhitespace(s, pos);
            if (pos[0] < s.length() && s.charAt(pos[0]) == ':') pos[0]++;
            skipWhitespace(s, pos);

            switch (key) {
                case "node_id":      input.nodeId = readString(s, pos); break;
                case "node_name":    input.nodeName = readString(s, pos); break;
                case "run_id":       input.runId = readString(s, pos); break;
                case "app_id":       input.appId = readString(s, pos); break;
                case "board_id":     input.boardId = readString(s, pos); break;
                case "user_id":      input.userId = readString(s, pos); break;
                case "stream_state": {
                    String v = readValue(s, pos);
                    input.streamState = "true".equals(v);
                    break;
                }
                case "log_level": {
                    String v = readValue(s, pos);
                    if (v.length() == 1 && v.charAt(0) >= '0' && v.charAt(0) <= '9') {
                        input.logLevel = v.charAt(0) - '0';
                    }
                    break;
                }
                case "inputs":
                    input.inputs.putAll(readMap(s, pos));
                    break;
                default:
                    readValue(s, pos);
            }
        }
        return input;
    }

    private static void skipWhitespace(String s, int[] pos) {
        while (pos[0] < s.length()) {
            char c = s.charAt(pos[0]);
            if (c != ' ' && c != '\t' && c != '\n' && c != '\r') break;
            pos[0]++;
        }
    }

    private static String readString(String s, int[] pos) {
        if (pos[0] >= s.length() || s.charAt(pos[0]) != '"') return "";
        pos[0]++;
        StringBuilder b = new StringBuilder();
        while (pos[0] < s.length() && s.charAt(pos[0]) != '"') {
            if (s.charAt(pos[0]) == '\\') {
                pos[0]++;
                if (pos[0] < s.length()) {
                    char esc = s.charAt(pos[0]);
                    switch (esc) {
                        case '"':  b.append('"'); break;
                        case '\\': b.append('\\'); break;
                        case 'n':  b.append('\n'); break;
                        case 'r':  b.append('\r'); break;
                        case 't':  b.append('\t'); break;
                        default:   b.append(esc);
                    }
                }
            } else {
                b.append(s.charAt(pos[0]));
            }
            pos[0]++;
        }
        if (pos[0] < s.length()) pos[0]++;
        return b.toString();
    }

    private static String readValue(String s, int[] pos) {
        skipWhitespace(s, pos);
        if (pos[0] >= s.length()) return "";
        char c = s.charAt(pos[0]);
        if (c == '"') {
            String v = readString(s, pos);
            return "\"" + v + "\"";
        }
        if (c == '{' || c == '[') {
            return readBracketed(s, pos, c == '{' ? '{' : '[', c == '{' ? '}' : ']');
        }
        int start = pos[0];
        while (pos[0] < s.length()) {
            char ch = s.charAt(pos[0]);
            if (ch == ',' || ch == '}' || ch == ']' || ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r') break;
            pos[0]++;
        }
        return s.substring(start, pos[0]);
    }

    private static String readBracketed(String s, int[] pos, char open, char close) {
        int depth = 0;
        int start = pos[0];
        while (pos[0] < s.length()) {
            char c = s.charAt(pos[0]);
            if (c == open) {
                depth++;
            } else if (c == close) {
                depth--;
                if (depth == 0) {
                    pos[0]++;
                    return s.substring(start, pos[0]);
                }
            } else if (c == '"') {
                pos[0]++;
                while (pos[0] < s.length() && s.charAt(pos[0]) != '"') {
                    if (s.charAt(pos[0]) == '\\') pos[0]++;
                    pos[0]++;
                }
            }
            pos[0]++;
        }
        return s.substring(start, pos[0]);
    }

    private static Map<String, String> readMap(String s, int[] pos) {
        Map<String, String> map = new HashMap<>();
        skipWhitespace(s, pos);
        if (pos[0] >= s.length() || s.charAt(pos[0]) != '{') {
            readValue(s, pos);
            return map;
        }
        pos[0]++;
        while (pos[0] < s.length()) {
            skipWhitespace(s, pos);
            if (pos[0] >= s.length() || s.charAt(pos[0]) == '}') {
                pos[0]++;
                break;
            }
            if (s.charAt(pos[0]) == ',') { pos[0]++; continue; }
            String k = readString(s, pos);
            skipWhitespace(s, pos);
            if (pos[0] < s.length() && s.charAt(pos[0]) == ':') pos[0]++;
            String v = readValue(s, pos);
            map.put(k, v);
        }
        return map;
    }

    public static final class ObjectBuilder {
        private final StringBuilder b = new StringBuilder();
        private boolean hasField;

        public ObjectBuilder() {
            b.append('{');
        }

        public ObjectBuilder field(String key, String value) {
            separator();
            b.append(quote(key)).append(':').append(quote(value));
            return this;
        }

        public ObjectBuilder rawField(String key, String rawValue) {
            separator();
            b.append(quote(key)).append(':').append(rawValue);
            return this;
        }

        public ObjectBuilder intField(String key, int value) {
            separator();
            b.append(quote(key)).append(':').append(value);
            return this;
        }

        public ObjectBuilder longField(String key, long value) {
            separator();
            b.append(quote(key)).append(':').append(value);
            return this;
        }

        public ObjectBuilder boolField(String key, boolean value) {
            separator();
            b.append(quote(key)).append(':').append(value ? "true" : "false");
            return this;
        }

        public ObjectBuilder doubleField(String key, double value) {
            separator();
            b.append(quote(key)).append(':').append(value);
            return this;
        }

        private void separator() {
            if (hasField) b.append(',');
            hasField = true;
        }

        public String build() {
            b.append('}');
            return b.toString();
        }
    }
}
